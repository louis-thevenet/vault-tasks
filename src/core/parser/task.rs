mod parse_completion;
mod parse_today;
mod parser_priorities;
mod parser_state;
mod parser_tags;
mod token;

use chrono::NaiveDateTime;
use parse_completion::parse_completion;
use parse_today::parse_today;
use parser_priorities::parse_priority;
use parser_state::parse_task_state;
use parser_tags::parse_tag;
use token::Token;
use tracing::error;
use winnow::{
    Parser, Result,
    combinator::{alt, fail, repeat},
    token::any,
};

use crate::core::{TasksConfig, date::Date, task::Task};

use super::{
    parser_date::{self},
    parser_time,
};

/// Parses a `Token` from an input string.FileEntry
fn parse_token(input: &mut &str, config: &TasksConfig) -> Result<Token> {
    alt((
        // |input: &mut &str| (parse_naive_date(input, config.use_american_format)),
        |input: &mut &str| {
            let date = parser_date::parse_naive_date(input, config.use_american_format)?;
            Ok(Token::DueDate(date))
        },
        |input: &mut &str| {
            let date = parser_time::parse_naive_time(input)?;
            Ok(Token::DueTime(date))
        },
        parse_tag,
        |input: &mut &str| parse_task_state(input, &config.task_state_markers),
        parse_priority,
        parse_completion,
        parse_today,
        |input: &mut &str| {
            let res = repeat(0.., any)
                .fold(String::new, |mut string, c| {
                    string.push(c);
                    string
                })
                .parse_next(input)?;
            Ok(Token::Name(res))
        },
    ))
    .parse_next(input)
}

/// Parses a `Task` from an input string. Filename must be specified to be added to the task.
///
/// # Errors
///
/// Will return an error if the task can't be parsed.
#[allow(clippy::module_name_repetitions)]
pub fn parse_task(input: &mut &str, filename: String, config: &TasksConfig) -> Result<Task> {
    let task_state = match parse_task_state(input, &config.task_state_markers)? {
        Token::State(state) => Ok(state),
        _ => fail(input),
    }?;

    let mut token_parser = |input: &mut &str| parse_token(input, config);

    let tokens = input
        .split_ascii_whitespace()
        .map(|token| token_parser.parse(token));

    let mut task = Task {
        state: task_state,
        filename,
        ..Default::default()
    };

    // Placeholders for a date and a time
    let mut due_date_opt = None;
    let mut due_time_opt = None;
    let mut name_vec = vec![]; // collects words that aren't tokens from the input string

    for token_res in tokens {
        match token_res {
            Ok(Token::DueDate(date)) => due_date_opt = Some(date),
            Ok(Token::DueTime(time)) => due_time_opt = Some(time),
            Ok(Token::Name(name)) => name_vec.push(name),
            Ok(Token::Priority(p)) => task.priority = p,
            Ok(Token::State(state)) => task.state = state,
            Ok(Token::Tag(tag)) => {
                if let Some(ref mut tags) = task.tags {
                    tags.push(tag);
                } else {
                    task.tags = Some(vec![tag]);
                }
            }
            Ok(Token::TodayFlag) => task.is_today = true,
            Ok(Token::CompletionPercentage(c)) => task.completion = Some(c),
            Err(error) => error!("Error: {error:?}"),
        }
    }

    if !name_vec.is_empty() {
        task.name = name_vec.join(" ");
    }
    let due_date = if let Some(due_date) = due_date_opt {
        match due_time_opt {
            Some(time) => Some(Date::DayTime(NaiveDateTime::new(due_date, time))),
            None => Some(Date::Day(due_date)),
        }
    } else {
        None
    };

    task.due_date = due_date;
    Ok(task)
}
#[cfg(test)]
mod test {

    use chrono::{Datelike, Days, NaiveDate, NaiveDateTime, NaiveTime};

    use crate::core::{
        TasksConfig,
        date::Date,
        parser::task::parse_task,
        task::{State, Task},
    };
    #[test]
    fn test_parse_task_no_description() {
        let mut input = "- [x] 10/15 task_name #done";
        let config = TasksConfig {
            use_american_format: true,
            ..Default::default()
        };
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let year = chrono::Local::now().year();
        let expected = Task {
            name: "task_name".to_string(),
            description: None,
            tags: Some(vec!["done".to_string()]),
            due_date: Some(Date::Day(NaiveDate::from_ymd_opt(year, 10, 15).unwrap())),
            priority: 0,
            state: State::Done,
            line_number: Some(1),
            ..Default::default()
        };
        assert_eq!(res, expected);
    }

    #[test]
    fn test_parse_task_only_state() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected = Task {
            subtasks: vec![],
            name: String::new(),
            description: None,
            tags: None,
            due_date: None,
            priority: 0,
            state: State::ToDo,
            line_number: Some(1),
            filename: String::new(),
            is_today: false,
            completion: None,
        };
        assert_eq!(res, expected);
    }
    #[test]
    fn test_parse_task_with_due_date_words() {
        let mut input = "- [ ] today 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_date = chrono::Local::now().date_naive();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_with_weekday() {
        let mut input = "- [ ] monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();

        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_with_weekday_this() {
        let mut input = "- [ ] this monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_with_weekday_next() {
        let mut input = "- [ ] next monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = Date::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, Some(expected_due_date));
    }

    #[test]
    fn test_parse_task_without_due_date() {
        let mut input = "- [ ] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_due_date = None;
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_invalid_state() {
        let mut input = "- [invalid] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_without_state() {
        let mut input = "task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_with_invalid_priority() {
        let mut input = "- [ ] task_name p-9";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.priority, 0);
    }

    #[test]
    fn test_parse_task_without_name() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.name, ""); // Default name is used when no name is provided
    }
    #[test]
    fn test_parse_task_with_today_flag() {
        let mut input = "- [ ] @t";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert!(res.is_today);
    }
}
