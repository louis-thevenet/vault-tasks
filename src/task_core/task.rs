use chrono::{NaiveDate, NaiveDateTime};
use color_eyre::{eyre::bail, Result};
use core::fmt;
use std::{
    fmt::Display,
    fs::{read_to_string, File},
    io::Write,
    path::PathBuf,
};
use tracing::{debug, info};

use crate::config::Config;

const STATE_TO_DO_EMOJI: &str = "‚ùå";
const STATE_DONE_EMOJI: &str = "‚úÖ";
pub const DUE_DATE_EMOJI: &str = "üìÖ";
pub const PRIORITY_EMOJI: &str = "‚ùó";
#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub enum State {
    Done,
    ToDo,
}
impl Display for State {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Done => write!(f, "{STATE_DONE_EMOJI}")?,
            Self::ToDo => write!(f, "{STATE_TO_DO_EMOJI}")?,
        }
        Ok(())
    }
}
#[derive(Debug, Hash, Eq, PartialEq, Clone)]
/// This type accounts for the case where the task has a due date but no exact due time
pub enum DueDate {
    NoDate,
    Day(NaiveDate),
    DayTime(NaiveDateTime),
}
impl Display for DueDate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Day(date) => write!(f, "{date}"),
            Self::DayTime(date) => write!(f, "{date}"),
            Self::NoDate => Ok(()),
        }
    }
}

impl DueDate {
    pub fn to_display_format(&self, not_american_format: bool) -> String {
        if matches!(self, Self::NoDate) {
            String::new()
        } else {
            format!(
                "{DUE_DATE_EMOJI} {}",
                self.to_string_format(not_american_format)
            )
        }
    }
    pub fn to_string_format(&self, not_american_format: bool) -> String {
        let format_date = if !not_american_format {
            "%Y/%m/%d"
        } else {
            "%d/%m/%Y"
        };
        let format_datetime = if !not_american_format {
            "%Y/%m/%d %T"
        } else {
            "%d/%m/%Y %T"
        };

        match self {
            Self::Day(date) => date.format(format_date).to_string(),
            Self::DayTime(date) => date.format(format_datetime).to_string(),
            Self::NoDate => String::new(),
        }
    }
}

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub struct Task {
    pub subtasks: Vec<Task>,
    pub description: Option<String>,
    pub due_date: DueDate,
    pub line_number: usize,
    pub name: String,
    pub priority: usize,
    pub state: State,
    pub tags: Option<Vec<String>>,
}

impl Default for Task {
    fn default() -> Self {
        Self {
            due_date: DueDate::NoDate,
            name: String::new(),
            priority: 0,
            state: State::ToDo,
            tags: None,
            description: None,
            line_number: 1,
            subtasks: vec![],
        }
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Name: \t -> {0}", self.name)?;

        match self.due_date {
            DueDate::NoDate => (),
            DueDate::DayTime(date) => writeln!(f, "Due: \t -> {date}")?,
            DueDate::Day(date) => writeln!(f, "Due: \t {date}")?,
        }

        writeln!(f, "State: \t -> {}", self.state)?;

        if self.priority > 0 {
            writeln!(f, "Prio.: \t -> {}", self.priority)?;
        }

        if let Some(description) = &self.description {
            writeln!(f, "Desc.:\t -> \"{description}\"")?;
        }
        if let Some(tags) = &self.tags {
            writeln!(f, "Tags: \t -> {tags:?}")?;
        }
        fmt::Result::Ok(())
    }
}
impl Task {
    fn get_fixed_attributes(&self, config: &Config, indent_length: usize) -> String {
        let indent = " ".repeat(indent_length);

        let state_str = match self.state {
            State::Done => "X",
            State::ToDo => " ",
        };

        let priority = if self.priority > 0 {
            format!("p{} ", self.priority)
        } else {
            String::new()
        };

        let mut due_date = self
            .due_date
            .to_string_format(!config.tasks_config.use_american_format);
        if !due_date.is_empty() {
            due_date.push(' ');
        }

        let tags_str = self.tags.as_ref().map_or_else(String::new, |tags| {
            tags.clone()
                .iter()
                .map(|t| format!("#{t}"))
                .collect::<Vec<String>>()
                .join(" ")
        });

        let res = format!(
            "{}- [{}] {} {}{}{}",
            indent, state_str, self.name, due_date, priority, tags_str
        );
        res.trim_end().to_string()
    }

    pub fn fix_task_attributes(&self, config: &Config, path: &PathBuf) -> Result<()> {
        let content = read_to_string(path.clone())?;
        let mut lines = content.split('\n').collect::<Vec<&str>>();

        if lines.len() < self.line_number - 1 {
            bail!(
                "Task's line number {} was greater than length of file {:?}",
                self.line_number,
                path
            );
        }

        let indent_length = lines[self.line_number - 1]
            .chars()
            .take_while(|c| c.is_whitespace())
            .count();

        let fixed_line = self.get_fixed_attributes(config, indent_length);

        if lines[self.line_number - 1] != fixed_line {
            debug!(
                "\nReplacing\n{}\nWith\n{}\n",
                lines[self.line_number - 1],
                self.get_fixed_attributes(config, indent_length,)
            );
            lines[self.line_number - 1] = &fixed_line;

            let mut file = File::create(path)?;
            file.write_all(lines.join("\n").as_bytes())?;

            info!("Wrote to {path:?} at line {}", self.line_number);
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use chrono::NaiveDate;

    use crate::{
        config::Config,
        task_core::task::{DueDate, State, Task},
    };

    #[test]
    fn test_fix_attributes() {
        let mut config = Config::default();
        config.tasks_config.use_american_format = true;
        let task = Task {
            due_date: DueDate::Day(NaiveDate::from_ymd_opt(2021, 12, 3).unwrap()),
            name: String::from("Test Task"),
            priority: 1,
            state: State::ToDo,
            tags: Some(vec![String::from("tag1"), String::from("tag2")]),
            description: Some(String::from("This is a test task.")),
            line_number: 2,
            subtasks: vec![],
        };
        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [ ] Test Task 2021/12/03 p1 #tag1 #tag2");
    }

    #[test]
    fn test_fix_attributes_with_no_date() {
        let config = Config::default();
        let task = Task {
            due_date: DueDate::NoDate,
            name: String::from("Test Task with No Date"),
            priority: 2,
            state: State::Done,
            tags: Some(vec![String::from("tag3")]),
            description: None,
            line_number: 3,
            subtasks: vec![],
        };

        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [X] Test Task with No Date p2 #tag3");
    }
}
