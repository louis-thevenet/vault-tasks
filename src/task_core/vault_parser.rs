use color_eyre::{eyre::bail, Result};
use std::{
    fs::{self, DirEntry},
    path::Path,
};
use tracing::{debug, info};

use crate::{config::Config, task_core::parser::parser_file_entry::ParserFileEntry};

use super::vault_data::VaultData;

pub struct VaultParser {
    config: Config,
}

impl VaultParser {
    pub const fn new(config: Config) -> Self {
        Self { config }
    }
    pub fn scan_vault(&self) -> Result<VaultData> {
        let mut tasks = VaultData::Directory(
            self.config
                .tasks_config
                .vault_path
                .to_str()
                .unwrap()
                .to_owned(),
            vec![],
        );
        info!("Scanning {:?}", self.config.tasks_config.vault_path);
        self.scan(&self.config.tasks_config.vault_path, &mut tasks)?;
        Ok(tasks)
    }

    fn scan(&self, path: &Path, tasks: &mut VaultData) -> Result<()> {
        if self.config.tasks_config.ignored.contains(&path.to_owned()) {
            debug!("Ignoring {path:?} (ignored list)");
            return Ok(());
        }

        let entries = path.read_dir()?;
        for entry_err in entries {
            let Ok(entry) = entry_err else { continue };
            let name = entry.file_name().into_string().unwrap();
            if !self.config.tasks_config.parse_dot_files && name.starts_with('.') {
                debug!("Ignoring {name:?} (dot file)");
                continue;
            }
            if self.config.tasks_config.ignored.contains(&entry.path()) {
                debug!("Ignoring {name:?} (ignored list)");
                continue;
            }

            if let VaultData::Directory(_, children) = tasks {
                if entry.path().is_dir() {
                    // recursive call for this subdir
                    let mut new_child = VaultData::Directory(
                        entry.file_name().to_str().unwrap().to_owned(),
                        vec![],
                    );

                    self.scan(&entry.path(), &mut new_child)?;

                    if let VaultData::Directory(_, c) = new_child.clone() {
                        if c.is_empty() {
                        } else {
                            children.push(new_child);
                        }
                    }
                } else if !std::path::Path::new(
                    &entry.file_name().into_string().unwrap_or_default(),
                )
                .extension()
                .map_or(false, |ext| ext.eq_ignore_ascii_case("md"))
                {
                    debug!("Ignoring {name:?} (not a .md file)");
                    continue;
                } else if let Some(file_tasks) = self.parse_file(&entry) {
                    children.push(file_tasks);
                }
            } else {
                bail!("Error while scanning directories, FileEntry was not a Directory");
            }
        }
        Ok(())
    }

    fn parse_file(&self, entry: &DirEntry) -> Option<VaultData> {
        debug!("Parsing {:?}", entry.file_name());
        let content = fs::read_to_string(entry.path()).unwrap_or_default();
        let parser = ParserFileEntry {
            config: &self.config,
        };

        parser.parse_file(entry.file_name().to_str().unwrap(), &content.as_str())
    }
}
