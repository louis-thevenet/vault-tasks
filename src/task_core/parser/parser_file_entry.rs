use std::{iter::Peekable, path::Path};

use color_eyre::{eyre::bail, Result};
use tracing::error;
use winnow::{
    ascii::{space0, space1},
    combinator::{alt, preceded, repeat},
    token::take_till,
    PResult, Parser,
};

use crate::task_core::{config::Config, file_entry::FileEntry, task::Task};

use super::task::parse_task;

enum FileToken {
    Header((String, usize)),    // Header, Depth
    Description(String, usize), // Content, indent length
    Task(Task, usize),          // Content, indent length
}

#[allow(clippy::module_name_repetitions)]
pub struct ParserFileEntry<'a> {
    pub config: &'a Config,
}

impl<'i> ParserFileEntry<'i> {
    fn parse_indent(input: &mut &str) -> PResult<usize> {
        let indent_length: String = repeat(1.., " ").parse_next(input)?;
        Ok(indent_length.len())
    }
    fn parse_task(&self, input: &mut &str) -> PResult<FileToken> {
        let indent_length = Self::parse_indent(input).unwrap_or(0);

        let mut task_parser = |input: &mut &str| parse_task(input, self.config);
        let task_res = task_parser.parse_next(input)?;
        Ok(FileToken::Task(task_res, indent_length))
    }
    fn parse_header(input: &mut &str) -> PResult<FileToken> {
        let header_depth: String = repeat(1.., "#").parse_next(input)?;
        let header_content = preceded(space0, take_till(1.., |c| c == '\n')).parse_next(input)?;

        Ok(FileToken::Header((
            header_content.to_string(),
            header_depth.len(),
        )))
    }
    fn parse_description(input: &mut &str) -> PResult<FileToken> {
        let indent_length = space1.map(|s: &str| s.len()).parse_next(input)?;
        let desc_content = take_till(1.., |c| c == '\n').parse_next(input)?;
        Ok(FileToken::Description(
            desc_content.to_string(),
            indent_length,
        ))
    }
    /// Inserts a `FileEntry` at the specific header `depth` in `file_entry`.
    fn insert_at(
        file_entry: &mut FileEntry,
        object: FileEntry,
        target_header_depth: usize,
        target_task_depth: usize,
    ) {
        fn insert_at_aux(
            file_entry: &mut FileEntry,
            object: FileEntry,
            current_header_depth: usize,
            target_header_depth: usize,
            current_task_depth: usize,
            target_task_depth: usize,
        ) {
            match file_entry {
                FileEntry::Header(_, header_children) => {
                    match (current_header_depth).cmp(&target_header_depth) {
                        std::cmp::Ordering::Greater => error!(
                            "bad call to `insert_at`, file_entry:{file_entry}\nobject:{object}"
                        ), // shouldn't happen
                        std::cmp::Ordering::Equal => {
                            // Found correct header level
                            if current_task_depth == target_task_depth {
                                header_children.push(object);
                            } else {
                                for child in header_children.iter_mut().rev() {
                                    if let FileEntry::Task(_, _) = child {
                                        return insert_at_aux(
                                            child,
                                            object,
                                            current_header_depth,
                                            target_header_depth,
                                            current_task_depth + 1,
                                            target_task_depth,
                                        );
                                    }
                                }
                            }
                        }
                        std::cmp::Ordering::Less => {
                            // Still haven't found correct header level, going deeper
                            for child in header_children.iter_mut().rev() {
                                if let FileEntry::Header(_, _) = child {
                                    insert_at_aux(
                                        child,
                                        object,
                                        current_header_depth + 1,
                                        target_header_depth,
                                        current_task_depth,
                                        target_task_depth,
                                    );
                                    return;
                                }
                            }
                            header_children.push(object);
                        }
                    }
                }
                FileEntry::Task(_task, subtasks) => {
                    if current_task_depth < target_task_depth {
                        for subtask in subtasks.iter_mut().rev() {
                            if let FileEntry::Task(_, _) = subtask {
                                return insert_at_aux(
                                    subtask,
                                    object,
                                    current_header_depth,
                                    target_header_depth,
                                    current_task_depth + 1,
                                    target_task_depth,
                                );
                            }
                        }
                    } else {
                        subtasks.push(object);
                    }
                }
            }
        }
        insert_at_aux(
            file_entry,
            object,
            0,
            target_header_depth,
            0,
            target_task_depth,
        );
    }

    /// Appends `desc` to the description of an existing `Task` in the `FileEntry`.
    fn append_description(
        file_entry: &mut FileEntry,
        desc: String,
        target_header_depth: usize,
        target_task_depth: usize,
    ) -> Result<()> {
        fn append_description_aux(
            file_entry: &mut FileEntry,
            desc: String,
            current_header_depth: usize,
            target_header_depth: usize,
            current_task_depth: usize,
            target_task_depth: usize,
        ) -> Result<()> {
            match file_entry {
                FileEntry::Header(_, header_children) => {
                    match current_header_depth.cmp(&target_header_depth) {
                        std::cmp::Ordering::Greater => panic!("bad call for desc"), // shouldn't happen
                        std::cmp::Ordering::Equal => {
                            // Found correct header level
                            for child in header_children.iter_mut().rev() {
                                if let FileEntry::Task(mut task, subtasks) = child.clone() {
                                    if current_task_depth == target_task_depth {
                                        match &mut task.description {
                                            Some(d) => {
                                                d.push('\n');
                                                d.push_str(&desc.clone());
                                            }
                                            None => task.description = Some(desc.clone()),
                                        }
                                        *child = FileEntry::Task(task, subtasks);
                                    } else {
                                        return append_description_aux(
                                            child,
                                            desc,
                                            current_header_depth,
                                            target_header_depth,
                                            current_task_depth + 1,
                                            target_task_depth,
                                        );
                                    }
                                }
                            }
                            Ok(())
                        }
                        std::cmp::Ordering::Less => {
                            // Going deeper in header levels
                            for child in header_children.iter_mut().rev() {
                                if let FileEntry::Header(_, _) = child {
                                    return append_description_aux(
                                        child,
                                        desc,
                                        current_header_depth + 1,
                                        target_header_depth,
                                        current_task_depth,
                                        target_task_depth,
                                    );
                                }
                            }
                            bail!("Failed to insert description: previous task not found");
                        }
                    }
                }
                FileEntry::Task(task, subtasks) => {
                    if current_task_depth == target_task_depth {
                        match &mut task.description {
                            Some(d) => {
                                d.push('\n');
                                d.push_str(&desc);
                                Ok(())
                            }
                            None => {
                                task.description = Some(desc.clone());
                                Ok(())
                            }
                        }
                    } else if let Some(last_task) = subtasks.last_mut() {
                        append_description_aux(
                            last_task,
                            desc,
                            current_header_depth,
                            target_header_depth,
                            current_task_depth + 1,
                            target_task_depth,
                        )
                    } else {
                        error!("Could not find description's parent task, indenting may be wrong. Closest task line number: {}",task.line_number);
                        bail!("Failed to insert description")
                    }
                }
            }
        }
        append_description_aux(
            file_entry,
            desc,
            0,
            target_header_depth,
            0,
            target_task_depth,
        )
    }

    /// Recursively parses the input file passed as a string.
    fn parse_file_aux<'a, I>(
        &self,
        mut input: Peekable<I>,
        file_entry: &mut FileEntry,
        header_depth: usize,
    ) where
        I: Iterator<Item = (usize, &'a str)>,
    {
        let mut parser = alt((
            Self::parse_header,
            |input: &mut &str| self.parse_task(input),
            Self::parse_description,
        ));

        let line_opt = input.next();
        if line_opt.is_none() {
            return;
        }

        let (line_number, mut line) = line_opt.unwrap();

        match parser.parse_next(&mut line) {
            Ok(FileToken::Task(mut task, indent_length)) => {
                task.line_number = line_number + 1; // line 1 was element 0 of iterator
                Self::insert_at(
                    file_entry,
                    FileEntry::Task(task, vec![]),
                    header_depth,
                    indent_length / self.config.indent_length,
                );
                self.parse_file_aux(input, file_entry, header_depth);
            }
            Ok(FileToken::Header((header, new_depth))) => {
                Self::insert_at(
                    file_entry,
                    FileEntry::Header(header, vec![]),
                    new_depth - 1,
                    0,
                );
                self.parse_file_aux(input, file_entry, new_depth);
            }
            Ok(FileToken::Description(description, indent_length)) => {
                if Self::append_description(
                    file_entry,
                    description.clone(),
                    header_depth,
                    indent_length / self.config.indent_length,
                )
                .is_err()
                {
                    error!("Failed to insert description {description}]");
                }
                self.parse_file_aux(input, file_entry, header_depth);
            }

            Err(_) => self.parse_file_aux(input, file_entry, header_depth),
        }
    }

    /// Removes any empty headers from a `FileEntry`
    fn clean_file_entry(file_entry: &mut FileEntry) -> Option<&FileEntry> {
        match file_entry {
            FileEntry::Header(_, children) => {
                let mut actual_children = vec![];
                for child in children.iter_mut() {
                    let mut child_clone = child.clone();
                    if Self::clean_file_entry(&mut child_clone).is_some() {
                        actual_children.push(child_clone);
                    }
                }
                *children = actual_children;
                if children.is_empty() {
                    None
                } else {
                    Some(file_entry)
                }
            }
            FileEntry::Task(_, _) => Some(file_entry),
        }
    }

    pub fn parse_file(&self, filename: &Path, input: &&str) -> Option<FileEntry> {
        let lines = input.split('\n');

        let filename = filename
            .strip_prefix(self.config.vault_path.clone())
            .unwrap();

        let mut res = FileEntry::Header(filename.to_str().unwrap().to_owned(), vec![]);
        self.parse_file_aux(lines.enumerate().peekable(), &mut res, 0);
        Self::clean_file_entry(&mut res).cloned()
    }
}
#[cfg(test)]
mod tests {

    use crate::task_core::{config::Config, file_entry::FileEntry, task::Task};

    use super::ParserFileEntry;

    #[test]
    fn test_with_useless_headers() {
        let input = r"# 1 useless
## 2 useless
### 3 useless

# 2 useful
### 3 useless
## 4 useful
- [ ] test
  test
  desc
"
        .split('\n')
        .enumerate()
        .peekable();

        let config = &Config::default();
        let mut res = FileEntry::Header("Test".to_string(), vec![]);
        let parser = ParserFileEntry { config };
        let expected = FileEntry::Header(
            "Test".to_string(),
            vec![
                FileEntry::Header(
                    "1 useless".to_string(),
                    vec![FileEntry::Header(
                        "2 useless".to_string(),
                        vec![FileEntry::Header("3 useless".to_string(), vec![])],
                    )],
                ),
                FileEntry::Header(
                    "2 useful".to_string(),
                    vec![
                        FileEntry::Header("3 useless".to_string(), vec![]),
                        FileEntry::Header(
                            "4 useful".to_string(),
                            vec![FileEntry::Task(
                                Task {
                                    name: "test".to_string(),
                                    line_number: 8,
                                    description: Some("test\ndesc".to_string()),
                                    ..Default::default()
                                },
                                vec![],
                            )],
                        ),
                    ],
                ),
            ],
        );
        parser.parse_file_aux(input, &mut res, 0);
        assert_eq!(res, expected);

        let expected_after_cleaning = FileEntry::Header(
            "Test".to_string(),
            vec![FileEntry::Header(
                "2 useful".to_string(),
                vec![FileEntry::Header(
                    "4 useful".to_string(),
                    vec![FileEntry::Task(
                        Task {
                            name: "test".to_string(),
                            line_number: 8,
                            description: Some("test\ndesc".to_string()),
                            ..Default::default()
                        },
                        vec![],
                    )],
                )],
            )],
        );
        ParserFileEntry::clean_file_entry(&mut res);
        assert_eq!(res, expected_after_cleaning);
    }
    #[test]
    fn test_simple_input() {
        let input = r"# 1 Header
- [ ] Task

## 2 Header
### 3 Header
- [ ] Task
- [ ] Task 2
## 2 Header 2
- [ ] Task
  Description

"
        .split('\n')
        .enumerate()
        .peekable();

        let config = &Config::default();
        let mut res = FileEntry::Header("Test".to_string(), vec![]);
        let parser = ParserFileEntry { config };
        let expected = FileEntry::Header(
            "Test".to_string(),
            vec![FileEntry::Header(
                "1 Header".to_string(),
                vec![
                    FileEntry::Task(
                        Task {
                            name: "Task".to_string(),
                            line_number: 2,
                            ..Default::default()
                        },
                        vec![],
                    ),
                    FileEntry::Header(
                        "2 Header".to_string(),
                        vec![FileEntry::Header(
                            "3 Header".to_string(),
                            vec![
                                FileEntry::Task(
                                    Task {
                                        name: "Task".to_string(),
                                        line_number: 6,
                                        ..Default::default()
                                    },
                                    vec![],
                                ),
                                FileEntry::Task(
                                    Task {
                                        name: "Task 2".to_string(),
                                        line_number: 7,
                                        ..Default::default()
                                    },
                                    vec![],
                                ),
                            ],
                        )],
                    ),
                    FileEntry::Header(
                        "2 Header 2".to_string(),
                        vec![FileEntry::Task(
                            Task {
                                name: "Task".to_string(),
                                line_number: 9,
                                description: Some("Description".to_string()),
                                ..Default::default()
                            },
                            vec![],
                        )],
                    ),
                ],
            )],
        );
        parser.parse_file_aux(input, &mut res, 0);
        assert_eq!(res, expected);
    }
    #[test]
    fn test_fake_description() {
        let input = r"# 1 Header
  test
- [ ] Task

## 2 Header
  test
"
        .split('\n')
        .enumerate()
        .peekable();

        let config = &Config::default();
        let mut res = FileEntry::Header("Test".to_string(), vec![]);
        let parser = ParserFileEntry { config };
        let expected = FileEntry::Header(
            "Test".to_string(),
            vec![FileEntry::Header(
                "1 Header".to_string(),
                vec![
                    FileEntry::Task(
                        Task {
                            name: "Task".to_string(),
                            line_number: 3,
                            ..Default::default()
                        },
                        vec![],
                    ),
                    FileEntry::Header("2 Header".to_string(), vec![]),
                ],
            )],
        );
        parser.parse_file_aux(input, &mut res, 0);
        assert_eq!(res, expected);
    }
}
