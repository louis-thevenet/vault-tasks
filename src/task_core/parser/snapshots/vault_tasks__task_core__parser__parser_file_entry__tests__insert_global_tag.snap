---
source: src/task_core/parser/parser_file_entry.rs
expression: res
---
0. Test
‾‾‾‾‾‾‾
	1. 1 Header
	‾‾‾‾‾‾‾‾‾‾‾
		Name: 	 -> Task
		State: 	 -> ❌
		Tags: 	 -> ["test"]
		
		2. 2 Header
		‾‾‾‾‾‾‾‾‾‾‾
			3. 3 Header
			‾‾‾‾‾‾‾‾‾‾‾
				Name: 	 -> Task
				State: 	 -> ❌
				Tags: 	 -> ["test"]
				
				Name: 	 -> Task 2
				State: 	 -> ❌
				Tags: 	 -> ["test"]
				
		2. 2 Header 2
		‾‾‾‾‾‾‾‾‾‾‾‾‾
			Name: 	 -> Task
			State: 	 -> ❌
			Desc.:	 -> "Description"
			Tags: 	 -> ["test"]
